Source/WebCore/page/SecurityOriginData.h

```diff
@@ -42,19 +49,33 @@ struct SecurityOriginData {
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
     }
+    explicit SecurityOriginData(OpaqueOriginIdentifier opaqueOriginIdentifier)
+        : protocol(emptyString())
+        , host(emptyString())
+        , port(std::nullopt)
+        , opaqueOriginIdentifier(opaqueOriginIdentifier)
+    {
+    }
+    // This constructor should only be used in cases where the origin must not be opaque. SecurityOriginData::fromURL should be used in most cases.
+    explicit SecurityOriginData(const URL& url)
+        : SecurityOriginData {
+            url.protocol().isNull() ? emptyString() : url.protocol().convertToASCIILowercase()
+            , url.host().isNull() ? emptyString() : url.host().convertToASCIILowercase()
+            , url.port()
+        }
+    {
+    }
     SecurityOriginData(WTF::HashTableDeletedValueType)
         : protocol(WTF::HashTableDeletedValue)
     {
     }
     
     WEBCORE_EXPORT static SecurityOriginData fromFrame(Frame*);
-    static SecurityOriginData fromURL(const URL& url)
+    WEBCORE_EXPORT static SecurityOriginData fromURL(const URL&);
+
+    static SecurityOriginData createOpaque()
```

Comment:
SecurityOriginData distinguishes the null state from the opaque state. The default constructor sets the strings to null, not the empty string. See this existing code:

```
    bool isNull() const
    {
        return protocol.isNull() && host.isNull() && port == std::nullopt;
    }
    bool isOpaque() const
    {
        return protocol == emptyString() && host == emptyString() && !port;
    }
```

As a result, I do not agree with your comment.

Reply:
For a SecurityOriginData created from default constructor, isOpaque() returns true before this patch, and returns false after this patch (which also modifies `isOpaque()`). Is this expected?

Reply:
I don't think isOpaque() returned true before this patch for default-constructed SecurityOriginData?

The protocol & host would be null, not the empty string and thus would fail the isOpaque() check, no?

Reply:
oh right, I misread this. It's actually three states:
1. url fields null, opaqueID invalid
2. url fields empty, opaqueID valid
3. url fields non-empty, opaqueID invalid